File: app/src/main/java/com/yanrou/dawnisland/DawnViewPager.java
package com.yanrou.dawnisland;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

public class DawnViewPager extends ViewPager {
    private float startX;
    private float startY;

    public DawnViewPager(@NonNull Context context) {
        super(context);
    }

    public DawnViewPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                getParent().requestDisallowInterceptTouchEvent(true);
                startX = ev.getX();
                startY = ev.getY();
                break;
            case MotionEvent.ACTION_MOVE:
                //来到新的坐标
                float endX = ev.getX();
                float endY = ev.getY();
                //计算偏移量
                float distanceX = endX - startX;
                float distanceY = endY - startY;
                //判断滑动方向
                //3是斜率的倒数
                if (Math.abs(distanceX) > (Math.abs(distanceY) * 3)) {
                    //水平方向滑动
//                   当滑动到ViewPager的第0个页面，并且是从左到右滑动
                    if (getCurrentItem() == 0 && distanceX > 0) {
                        getParent().requestDisallowInterceptTouchEvent(false);
                    }

//                  当滑动到ViewPager的最后一个页面，并且是从右到左滑动
//
                    else if ((getCurrentItem() == (getAdapter().getCount() - 1)) && distanceX < 0) {
                        getParent().requestDisallowInterceptTouchEvent(true);
                    }
                    //其他,中间部分
                    else {
                        getParent().requestDisallowInterceptTouchEvent(true);
                    }
                } else {
                    //竖直方向滑动
                    getParent().requestDisallowInterceptTouchEvent(true);
                }
                break;
            case MotionEvent.ACTION_UP:
                break;
            default:
                break;
        }
        return super.dispatchTouchEvent(ev);
    }
}

File: app/src/main/java/com/yanrou/dawnisland/DawnViewPager.kt
package com.yanrou.dawnisland

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.viewpager.widget.ViewPager
import timber.log.Timber
import java.lang.Math.abs

class DawnViewPager : ViewPager {
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    private var swipeLastX: Float = 0f
    private var swipeStartX: Float = 0f
    private var swipeCancel = false

    private var xDistance: Float = 0f
    private var yDistance: Float = 0f
    private var xLast: Float = 0f
    private var yLast: Float = 0f
    private val touchSlop: Int = android.view.ViewConfiguration.get(context).scaledTouchSlop
    private val swipeSlop: Int = 48
    private var isTouching = false

    var drawerListener: (() -> Unit)? = null

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        var handled = false
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                swipeStartX = event.x
                swipeLastX = 0f
                swipeCancel = false

                yDistance = 0f
                xDistance = yDistance
                xLast = event.x
                yLast = event.y
                return super.onInterceptTouchEvent(event)
            }
            MotionEvent.ACTION_MOVE -> {
                val curX = event.x
                val curY = event.y
                xDistance = abs(curX - xLast)
                yDistance = abs(curY - yLast)
                xLast = curX
                yLast = curY
                handled = abs(xDistance) >= touchSlop && yDistance * 4 < xDistance * 3
            }
        }

        // Fixed: IllegalArgumentException: pointerIndex out of range
        return try {
            handled || super.onInterceptTouchEvent(event)
        } catch (e: Throwable) {
            e.printStackTrace()

            return handled
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        if (currentItem == 0) {
            val action = ev.action
            val x = ev.x
            when (action) {
                MotionEvent.ACTION_DOWN -> isTouching = true
                MotionEvent.ACTION_MOVE -> {
                    isTouching = true
                    Timber.d("x: $x, lastX: $swipeLastX")
                    // 如果手指往回收，则取消开启侧栏
                    if (x < swipeLastX) {
                        swipeCancel = true
                    }
                    swipeLastX = x
                    // writer.drawerDelegate.drawer.setDrawerToOffset(x - swipeStartX)
                    // writer.drawerDelegate.drawer.invalidate()
                }
                MotionEvent.ACTION_UP -> {
                    if (!swipeCancel && (x - swipeStartX) > swipeSlop) {
                        drawerListener?.let { it() }
                    }
                    isTouching = false
                }
                MotionEvent.ACTION_CANCEL -> isTouching = false
            }
        }
        return super.onTouchEvent(ev)
    }

}